specs/opcode/31BALANCE.md

```diff
@@ -0,0 +1,49 @@
+# BALANCE opcode
+
+## Procedure
+
+The `BALANCE` opcode gets balance of the given account.
+
+## EVM behaviour
+
+The `BALANCE` opcode pops `address` (20 bytes of data) off the stack and pushes
+the balance of the corresponding account onto the stack. If the given account
+doesn't exist, then it will push 0 onto the stack instead.
+
+## Circuit behaviour
+
+1. Construct call context table in rw table.
+2. Do bus-mapping lookup for stack read, call context read and account read
+   operations.
+3. Do bus-mapping lookup for transaction access list write and stack write
+   operations.
+
+## Constraints
+
+1. opId = 0x31
+2. State transition:
+   - gc + 7 (1 stack read, 1 stack write, 3 call context reads, 1 account reads,
+     1 transaction access list write)
+   - stack_pointer + 0 (one pop and one push)
+   - pc + 1
+   - gas:
+     - the accessed `address` is warm: GAS_COST_WARM_ACCESS
+     - the accessed `address` is cold: GAS_COST_ACCOUNT_COLD_ACCESS
+3. Lookups: 7
+   - `address` is popped from the stack.
```

Comment:
I don't think this counts as a lookup. The lookup is done using this item to other tables.

Reply:
Hi @CPerezz, I tried to reference other spec documents and the corresponding code. It seems that the stack arguments are also counted as bus-mapping lookups as:

[ADD-SUB spec for lookups](https://github.com/privacy-scaling-explorations/zkevm-specs/blob/master/specs/opcode/01ADD_03SUB.md#constraints)
[ADD-SUB code for rw_counter](https://github.com/privacy-scaling-explorations/zkevm-specs/blob/master/src/zkevm_specs/evm/execution/add_sub.py#L21)

[EXTCODEHASH spec for lookups](https://github.com/privacy-scaling-explorations/zkevm-specs/blob/master/specs/opcode/3fEXTCODEHASH.md#constraints)
[EXTCODEHASH code for rw_counter](https://github.com/privacy-scaling-explorations/zkevm-specs/blob/master/src/zkevm_specs/evm/execution/extcodehash.py#L33)

I updated this part as:

```
...
3. Lookups: 7 busmapping lookups
   - `address` is at top of the stack.
   ...
   - The BALANCE result is at the new top of the stack.
...
```

Please help comment if I was wrong. Thanks.